Select Topbar

supergibbs opened the pull request:
Updated Topbar with a select option if `urls` is passed
I took @shockey's code from https://github.com/swagger-api/swagger-ui/issues/3103#issuecomment-302623899 and updated it to work with lastest master branch as an option instead of separate plugin.

Seems like a common request. Happy to work with you if this isn't the direction you want to take.

Adresses:
https://github.com/swagger-api/swagger-ui/issues/3103
https://github.com/swagger-api/swagger-ui/issues/2891

webron commented:
Thanks for the contribution, @supergibbs.

@shockey - thoughts about a config vs plugin for this one?

shockey commented:
@webron, I'm always inclined to say plugin, but I think this integrates deeply enough with Topbar (which is itself a plugin), and is widely applicable enough, to be okay as a direct config option.

webron commented:
@supergibbs so... tried running your changes (I'm letting @shockey do the real hard work of actually reviewing the code), and it doesn't quite work as I would expect it to? I'm getting an empty drop down, with the number of (empty) options. Choosing any just brings up an error message, and if `url` is defined as well, something even weirder happens. Can you look into that please?

supergibbs commented:
Here is how it's suppose to work:

- `urls` config is an array of spec objects that are just `{name: name, url: url}`
- if you also supply a `url` config to the full spec it'll add a link to it

We broke our big API into small chunks but have the full spec mostly for the client generator and some prefer seeing it all at once.

```
window.onload = function () {
  var apis = [
    {
      url: 'http://specs.com/spec1.yaml',
      name: 'Spec1'
    },
    {
      url: 'http://specs.com/spec1.yaml',
      name: 'Spec2'
    },
    {
      url: 'http://specs.com/spec1.yaml',
      name: 'Spec3'
    }
  ];

  const ui = SwaggerUIBundle({
    urls: apis, // array of spec objects
    url: 'http://specs.com/swagger.yaml', //Optional full spec
    dom_id: '#swagger-ui',
    validatorUrl: null,
    presets: [
      SwaggerUIBundle.presets.apis,
      SwaggerUIStandalonePreset
    ],
    plugins: [
      SwaggerUIBundle.plugins.DownloadUrl
    ],
    layout: "StandaloneLayout"
  })

  window.ui = ui
}
```

webron commented:
Thanks. That means that the README needs to be updated, because nobody could guess that's the expected structure.

Functionally, I think it's fine, and don't mind including it, especially if it helps various users. The styling is off, and @Minasokoni would need to provide his feedback on it (he might make changes to the styling before merging).

I find the mixed url/urls functionality unnecessary and would prefer the two to be mutually exclusive. If you'd like to have the full spec, just include it as the first or last entry in the APIs. By using the current solution, you're overloading the meaning of the `url` parameter as in one case it renders as a link and the other as an input text box. There's really no added functionality here that cannot be accomplished by adding it to the `urls`. Do you have any reasons to have it as-is that I didn't think of?

supergibbs commented:
Fair enough, I'll update with some more explicit configs and document it in the README. Ya, the full spec can just be added to the select. For precedence, which should be used if both `url` and `urls` are provided?

How is the styling off? I am just using the default styling. I set a class since @shockey did in his example but don't think it's doing anything. The flex layout stuff is weird but I'll try and make that simpler.

webron commented:
I'd say `url` is the basic, and things like `urls` or `spec` take over that. In terms of order - `url` < `urls` < `spec`.

The drop down is aligned to the right and not centered, and the size seems to be dynamic based on the length of the longest text. My take is that it should be more aligned with the `url` input box and have a fixed length. I'd much rather get @Minasokoni's input on that though.

I don't know if/how this feature can be tested (and it probably should be). @shockey any thoughts?

Thanks for accommodating, and thanks for taking the time to write up this PR. We appreciate the contribution.

supergibbs commented:
OK, I removed the "Full Spec" link. Personally I feel like the dropdown should be right aligned. The current input/button is left aligned but it's also 100% width. I cleaned up the CSS to simplify the component and tried to match the styling better. I set a min-width so it'll be fairly consistent but it'll grow as needed if there are long names. It's all in the scss file now so @Minasokoni can make the final call. I updated the README too.

Minasokoni commented:
@supergibbs thanks, I'll check it out

supergibbs commented:
Cool, I like the label but it's inheriting some styles from `.swagger-ui label`. There is a bottom margin and the font size is pretty small. Is that intended?

supergibbs commented:
Visual: 
![image](https://user-images.githubusercontent.com/1106331/27399670-6b4575a0-5672-11e7-8b04-3b17e13b05f8.png)

(bottom one has `margin: 0, font-size: 16px` added to `.select-label`)

shockey commented on README.md:
```diff
@@ -129,6 +129,7 @@ If you'd like to use the bundle files via npm, check out the [`swagger-ui-dist`
 Parameter Name | Description
 --- | ---
 url | The url pointing to API definition (normally `swagger.json` or `swagger.yaml`).
```
Should add "Superseded by `urls` parameter", or something to that effect.

supergibbs replied:
👍

webron commented:
@supergibbs let me know when you're done with your changes, and I'll review the functionality (and visuals) again. We're having a release tomorrow, would love to try and get it out with it.

supergibbs commented:
I'm all set, functionally. I just updated the README per @shockey's request. Not sure if the styling in final though (see above comments). I can make the change I suggested but waiting for sign off from @Minasokoni

shockey commented on src/plugins/topbar/topbar.jsx:
```diff
@@ -36,28 +72,52 @@ export default class Topbar extends React.Component {
+      })
+
+      control.push(
+        <label className="select-label" htmlFor="select"><span>Select a URL</span>
```
"Select a URL" feels strange since we're displaying names.

Minasokoni replied:
I didn't even realize that  😞

shockey replied:
Every dog has its day, @Minasokoni 😄

shockey commented on src/plugins/topbar/topbar.jsx:
```diff
@@ -36,28 +72,52 @@ export default class Topbar extends React.Component {
+    if(urls) {
+      let rows = []
+      urls.forEach((link, i) => {
+        rows.push(<option key={i} value={link.url} selected={i===this.state.selectedIndex}>{link.name}</option>)
```
This causes a React warning, please fix: 

> Warning: Use the `defaultValue` or `value` props on <select> instead of setting `selected` on <option>.

supergibbs replied:
Fixed

shockey commented on src/plugins/topbar/topbar.jsx:
```diff
@@ -19,14 +20,49 @@ export default class Topbar extends React.Component {
     e.preventDefault()
   }
 
+  componentWillMount() {
```
It's not clear to me what this method handles (with regards to the feature), or why it's needed. I removed it and my Topbar continued to select the first spec in `urls`, as I would expect.

It appears to read the browser query string for a name property, which is strange for two reasons:

- name isn't documented in the README as a config parameter
- components shouldn't read query strings directly, because it bypasses the YAML/constructor/qstring config inheritance system exposed via getConfigs.

Let me know if I'm reading this wrong.

supergibbs replied:
This is so you can link to a spec with `?name=Spec%20Name`. I didn't notice all the configs get merged. I've updated to use getConfigs and documented in the README better.

shockey commented:
@supergibbs, please see my remarks above 😄 

If you don't have the bandwidth to make the changes soon, let me know, and I'll try to get this PR across the line.

supergibbs commented:
Was out of town over the weekend but I'll make those changes soon.

On Mon, Jun 26, 2017 at 3:20 PM shockey <notifications@github.com> wrote:

> @supergibbs <https://github.com/supergibbs>, please see my remarks above
> 😄
>
> If you don't have the bandwidth to make the changes soon, let me know, and
> I'll try to get this PR across the line.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/swagger-api/swagger-ui/pull/3261#issuecomment-311198318>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABDhm9uN3auaFdhfwDzVAFb-N57p6kiiks5sIC6bgaJpZM4N-fpc>
> .
>

shockey commented:
Made some changes myself:

- `name` parameter is now `urls.primaryName`. We decided that using `.`s to denote subordinate parameters was the way to go, since we need a flat structure to support configuration via query string.
- Fixed a bug where the select menu wouldn't update when a new URL was selected.

shockey commented:
Thanks for all the work here, @supergibbs! We really appreciate the feature and the time you spent addressing our requests.

This code should go live later this week when we do our release.

supergibbs commented:
Happy to help and excited to use the next release!

webron commented:
@supergibbs Thanks again for the efforts and keeping up with our crazy thoughts ;) I'm sure many users would appreciate this feature. 

Looking forward to additional contributions should you find the time!

lukastosic commented:
Hi @supergibbs, this is really nice and works fine as you described in comment above (with `apis` variable) .. but can it work with getting array of name/url from json file ?
I tried making json file and putting its address to `urls` but it fails ...

supergibbs commented:
No it doesn't at the moment. You should be able to make a GET request to your JSON file though and read it into a variable.

hyfather commented:
Thanks for this @supergibbs, very helpful.
Quick question -- is there a way to allow a user to select a spec from the drop-down *or* have them enter their own spec URL if they don't find what they're looking for?

supergibbs commented:
No, I don't really see the use case but if you have one, feel free to make a PR :)

janholinka commented:
Hi,
is it possible to set urls property in docker-compose.yml as an environment variable? 

Thanks

shockey commented:
@janholinka, not currently. Adding that option is included in https://github.com/swagger-api/swagger-ui/issues/3902.

thescouser89 commented:
Hi,
Is it possible to set different authentications for the different urls?

shockey commented:
@thescouser89, you can use a `requestInterceptor` to apply different authentications to the various URLs you want to display!

thescouser89 commented:
@shockey Awesome! Will try! :) Thanks for the fast reply!

heyakyra commented:
Despite providing a `urls` config which can be accessed by the text input successfully, the topbar is rendering text input rather than a drop down. Any ideas on how to fix this?

In case it's helpful:
```
JSON.stringify(versions)
"{\"swaggerUi\":{\"version\":\"3.48.0\",\"gitRevision\":\"g38c6e12\",\"gitDirty\":true,\"buildTimestamp\":\"Thu, 29 Apr 2021 22:09:45 GMT\",\"machine\":\"ip-172-31-21-173\"}}"
```

And
```
      const ui = SwaggerUIBundle({
        urls: [
          { url: "./openapi/example.yaml", name: "Example API" },
          { url: "./openapi/example2.yaml", name: "Example API 2" }
        ],
        dom_id: '#swagger-ui',
        deepLinking: true,
        presets: [
          SwaggerUIBundle.presets.apis,
          SwaggerUIStandalonePreset
        ],
        plugins: [
          SwaggerUIBundle.plugins.DownloadUrl
        ],
        layout: "StandaloneLayout"
      })
```
